### PR TITLE
Add error API

### PR DESCRIPTION
**Pull request by luebbers**
Implements OPAE error API.

Exposes a property on the number of supported error registers.
Users can get information about the errors using fpgaGetErrorInfo(),
read errors using fpgaReadError() and clear errors using
fpgaClearError() and fpgaClearAllErrors().

Error sysfs nodes are dynamically parsed from the respective
sysfs directories (some are excluded based on a blacklist of
names). If there is a "clear" file, the "errors" file in sysfs is
assumed do be "clearable" (so far only true for PORT errors).

### CODE REVIEW

File: libopae/src/error.c [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+			n--;
+			break;
+		}
+		strncpy_s(new_entry->info.name, FPGA_ERROR_NAME_MAX, de->d_name, FILENAME_MAX);
+		strncpy_s(new_entry->error_file, SYSFS_PATH_MAX, basedir, FILENAME_MAX);
+		new_entry->next = NULL;
+		// see if error can be cleared (currently only errors called "errors" corresponding
+		// to a "clear" file can)
+		new_entry->info.can_clear = false;
+		if (strcmp(de->d_name, "errors") == 0) {    // FIXME: SAFE
```
Comment (r-rojo) [inline libopae/src/error.c]: Will this `FIXME` be addressed in another pull request?

File: libopae/src/error.c [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+		if (!new_entry) {
+			FPGA_MSG("can't allocate memory");
+			n--;
+			break;
+		}
+		strncpy_s(new_entry->info.name, FPGA_ERROR_NAME_MAX, de->d_name, FILENAME_MAX);
+		strncpy_s(new_entry->error_file, SYSFS_PATH_MAX, basedir, FILENAME_MAX);
+		new_entry->next = NULL;
+		// see if error can be cleared (currently only errors called "errors" corresponding
+		// to a "clear" file can)
```
Comment (r-rojo) [inline libopae/src/error.c]: So clearing errors will only be supported for errors that are cleared by writing to the `clear` file?
Will we eventually support clearing other types of errors?

File: libopae/src/error.c [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+}
+
+/* files and directories to ignore when looking for errors */
+const char *errors_exclude[] = {
+	"revision",
+	"uevent",
+	"power",
+	"clear"
+};
+#define NUM_ERRORS_EXCLUDE (sizeof(errors_exclude) / sizeof(char *))
```
Comment (r-rojo) [inline libopae/src/error.c]: Wouldn't it be better to more explicit on what nodes we operate on and whitelist the files? 
This would make it easier to determine how to clear errors too (without relying on the existence of a "clear" file).

File: libopae/src/error.c (hunk 1 above)
Comment (luebbers) [reply libopae/src/error.c]: No - I'll address that in this PR; it'll get shot down by static analysis otherwise.

File: libopae/src/error.c (hunk 2 above)
Comment (luebbers) [reply libopae/src/error.c]: We should. This follows what the mock driver models, but that should be expanded.

File: libopae/src/error.c (hunk 3 above)
Comment (luebbers) [reply libopae/src/error.c]: I was looking for something that minimizes maintenance if people start adding additional errors. Perhaps being more explicit about the naming of clearable errors would help (like "errors" and "errors_clear")?

**Comment by luebbers** [comment]
There are a number of limitations with the current approach in terms of performance and thread safety. I'll close this PR and open a new one.

File: libopae/src/token_list.c [hunk 4]
```diff
@@ -85,6 +86,11 @@ struct _fpga_token *token_add(const char *sysfspath, const char *devpath)
 		return NULL;
 	}
 
+	/* populate error list */
+	char errpath[SYSFS_PATH_MAX];
+	snprintf_s_s(errpath, SYSFS_PATH_MAX, "%s/errors", sysfspath);
```
Comment (r-rojo) [inline libopae/src/token_list.c]: what about fme errors? those are found in the subdirectory `errors/fme-errors/errors` with their clear file being `errors/fme-errors/clear`

Comment (r-rojo) [reply libopae/src/token_list.c]: Nevermind, I forgot this function recurses subdirectories

**Comment by lzhan55** [comment]
Just have a general question about error registers: are they part of the FME error registers located in the error DFH of FME? Or, each AFU has its own error registers in AFU's MMIO? Is there any error register related to AFUs?

**Comment by luebbers** [comment]
Added support for clearing FME errors and rebased on top of master. Please review again. Thanks! :)

File: libopae/src/error.c [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+	}
+
+	while ((de = readdir(dir))) {
+		// skip hidden ('.*') files (includes "." and "..")
+		if (de->d_name[0] == '.')
+			continue;
+
+		// skip names on blacklist
+		for (i = 0; i < NUM_ERRORS_EXCLUDE; i++) {
+			if (strcmp(de->d_name, errors_exclude[i]) == 0) {  // FIXME: SAFE?
```
Comment (r-rojo) [inline libopae/src/error.c]: You should probably use the safestring version of this

File: libopae/src/error.c [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+				break;
+			}
+		}
+		if (i < NUM_ERRORS_EXCLUDE)
+			continue;
+
+		// build absolute path
+		strncpy(basedir + len, de->d_name, FILENAME_MAX - len);
```
Comment (r-rojo) [inline libopae/src/error.c]: You should probably use one of the safestring variants of this.

File: libopae/src/error.c [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+
+		// append error info to list
+		struct error_list *new_entry = malloc(sizeof(struct error_list));
+		if (!new_entry) {
+			FPGA_MSG("can't allocate memory");
+			n--;
+			break;
+		}
+		strncpy_s(new_entry->info.name, FPGA_ERROR_NAME_MAX, de->d_name, FILENAME_MAX);
+		strncpy_s(new_entry->error_file, SYSFS_PATH_MAX, basedir, FILENAME_MAX);
```
Comment (r-rojo) [inline libopae/src/error.c]: Should we check for errors (by looking at the return value of `strncpy_s`)?

File: libopae/src/error.c [hunk 8]
```diff
@@ -0,0 +1,328 @@
...
+		new_entry->next = NULL;
+		// Errors can be cleared:
+		//   * if the name is "errors" and there is a file called "clear" (generic case), OR
+		//   * if the name is in the "errors_clearable" table
+		new_entry->info.can_clear = false;
+		if (strcmp(de->d_name, "errors") == 0) {
```
Comment (r-rojo) [inline libopae/src/error.c]: safestring version of strcmp?

File: libopae/src/error.c [hunk 9]
```diff
@@ -0,0 +1,328 @@
...
+			strncpy_s(basedir + len, FILENAME_MAX - len, "clear", sizeof("clear"));
+			// try accessing clear file
+			if (lstat(basedir, &st) != -1) {
+				new_entry->info.can_clear = true;
+				strncpy_s(new_entry->clear_file, SYSFS_PATH_MAX, basedir, FILENAME_MAX);
+			}
+		} else {
+			for (i = 0; i < NUM_ERRORS_CLEARABLE; i++) {
+				if (strcmp(de->d_name, errors_clearable[i]) == 0) {
```
Comment (r-rojo) [inline libopae/src/error.c]: safestring version of strcmp?

File: libopae/src/error.c [hunk 10]
```diff
@@ -0,0 +1,328 @@
...
+		// Errors can be cleared:
+		//   * if the name is "errors" and there is a file called "clear" (generic case), OR
+		//   * if the name is in the "errors_clearable" table
+		new_entry->info.can_clear = false;
+		if (strcmp(de->d_name, "errors") == 0) {
+			strncpy_s(basedir + len, FILENAME_MAX - len, "clear", sizeof("clear"));
+			// try accessing clear file
+			if (lstat(basedir, &st) != -1) {
+				new_entry->info.can_clear = true;
+				strncpy_s(new_entry->clear_file, SYSFS_PATH_MAX, basedir, FILENAME_MAX);
```
Comment (r-rojo) [inline libopae/src/error.c]: Check return value of `strncpy_s`?

File: libopae/src/error.c [hunk 11]
```diff
@@ -0,0 +1,328 @@
...
+			}
+		} else {
+			for (i = 0; i < NUM_ERRORS_CLEARABLE; i++) {
+				if (strcmp(de->d_name, errors_clearable[i]) == 0) {
+					strncpy_s(basedir + len, FILENAME_MAX - len, de->d_name, FILENAME_MAX);
```
Comment (r-rojo) [inline libopae/src/error.c]: Check return value of `strncpy_s`?

File: libopae/src/error.c [hunk 12]
```diff
@@ -0,0 +1,328 @@
...
+					// try accessing clear file
+					if (lstat(basedir, &st) != -1) {
+						new_entry->info.can_clear = true;
+						strncpy_s(new_entry->clear_file, SYSFS_PATH_MAX, basedir, FILENAME_MAX);
```
Comment (r-rojo) [inline libopae/src/error.c]: Check return value of `strncpy_s`?

File: libopae/src/error.c [hunk 13]
```diff
@@ -0,0 +1,328 @@
...
+					}
+				}
+			}
+		}
+
+		if (!new_entry->info.can_clear) {
+			memset_s(new_entry->clear_file, sizeof(new_entry->clear_file), 0);
```
Comment (r-rojo) [inline libopae/src/error.c]: Check return value of `memset_s`?

File: libopae/src/error.c [hunk 14]
```diff
@@ -0,0 +1,328 @@
...
+	uint32_t n = 0;
+	unsigned int i;
+	struct error_list **el = list;
+
+	if (len+1 > FILENAME_MAX) {
+		FPGA_MSG("path too long");
+		return 0;
+	}
+
+	strncpy(basedir, path, FILENAME_MAX-1);
```
Comment (r-rojo) [inline libopae/src/error.c]: safestring version?

File: libopae/src/error.c [hunk 15]
```diff
@@ -0,0 +1,328 @@
...
+#define NUM_ERRORS_CLEARABLE (sizeof(errors_clearable) / sizeof(char *))
+
+/* Walks the given directory and adds error entries to `list`.
+ * This function is called during enumeration when adding tokens to
+ * the global tokens list. When tokens are cloned, their error
+ * lists are only shallowly copied (which works because errors of
+ * a token never change).
+ * returns the number of error entries added to `list` */
+uint32_t __FIXME_MAKE_VISIBLE__
+build_error_list(const char *path, struct error_list **list)
```
Comment (r-rojo) [inline libopae/src/error.c]: Just out of curiosity, what happens if we call this twice with the same list? Will it duplicate the list?

File: libopae/src/error.c [hunk 16]
```diff
@@ -0,0 +1,328 @@
...
+			n--;
+			break;
+		}
+		strncpy_s(new_entry->info.name, FPGA_ERROR_NAME_MAX, de->d_name, FILENAME_MAX);
+		strncpy_s(new_entry->error_file, SYSFS_PATH_MAX, basedir, FILENAME_MAX);
+		new_entry->next = NULL;
+		// Errors can be cleared:
+		//   * if the name is "errors" and there is a file called "clear" (generic case), OR
+		//   * if the name is in the "errors_clearable" table
+		new_entry->info.can_clear = false;
```
Comment (r-rojo) [inline libopae/src/error.c]: Should the routine to check if an error file is clearable be refactored into its own function?

File: libopae/src/error_int.h [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+struct error_list {
+	struct fpga_error_info info;
+	struct error_list *next;
+	char error_file[SYSFS_PATH_MAX];
+	char clear_file[SYSFS_PATH_MAX];
+};
```
Comment (r-rojo) [inline libopae/src/error_int.h]: Have you thought about wrapping this in an outer struct with metadata (like the error count)?
It could look something like (and maybe more interesting metadata):
```C
struct error_info {
        struct error_list *errors;
        size_t count;
};
```

File: libopae/src/error.c (hunk 5 above)
Comment (luebbers) [reply libopae/src/error.c]: strcmp() is not a critical/banned function. But I should remove the comment.

File: libopae/src/error.c (hunk 15 above)
Comment (luebbers) [reply libopae/src/error.c]: Yes, it would append all errors it finds in `path` to the list `list`. It actually calls itself recursively. I'll add a comment.

File: libopae/src/error.c (hunk 7 above)
Comment (luebbers) [reply libopae/src/error.c]: Yes, we should. I'll add it.

File: libopae/src/error.c (hunk 16 above)
Comment (luebbers) [reply libopae/src/error.c]: Good idea. I'll look into it.

File: libopae/src/error.c (hunk 8 above)
Comment (luebbers) [reply libopae/src/error.c]: `strcmp()` is not a critical/banned function.

File: libopae/src/error.c (hunk 10 above)
Comment (luebbers) [reply libopae/src/error.c]: Yep, see above.

File: libopae/src/error.c (hunk 9 above)
Comment (luebbers) [reply libopae/src/error.c]: See above (not critical).

File: libopae/src/error.c (hunk 11 above)
Comment (luebbers) [reply libopae/src/error.c]: See above.

File: libopae/src/error.c (hunk 12 above)
Comment (luebbers) [reply libopae/src/error.c]: ...

File: libopae/src/error.c (hunk 13 above)
Comment (luebbers) [reply libopae/src/error.c]: Good idea, will do.

File: libopae/src/error_int.h (hunk 17 above)
Comment (luebbers) [reply libopae/src/error_int.h]: `error_info` is the structure handled (and accessed directly) by users of the API. I'd rather not expose implementation-specific information (like sysfs paths) to them, since they'll start relying on that and it will break as soon as (for example) there's not sysfs entry for an error.

File: libopae/src/error.c (hunk 5 above)
Comment (luebbers) [reply libopae/src/error.c]: done

File: libopae/src/error.c (hunk 6 above)
Comment (luebbers) [reply libopae/src/error.c]: done

File: libopae/src/error.c (hunk 7 above)
Comment (luebbers) [reply libopae/src/error.c]: done

File: libopae/src/error.c (hunk 16 above)
Comment (r-rojo) [reply libopae/src/error.c]: Ok, it can be a TODO...
